content/blog/graphql-as-a-gateway/index.md

```diff
@@ -0,0 +1,200 @@
+
+## ⚡ GraphQL as your API Gateway
+
+High hopes that you are convinced on trying GraphQL 🙌 Implementing a GraphQL server is not complicated and there are many guides in the web talking about it. The official website [has a list with many server frameworks and libraries](https://graphql.org/code) (comes in many flavours).
+
+> Even before going full micro-services and using it as a real API Gateway, companies can easily develop everything on top of a GraphQL monolith ~~wait, don' leave yet~~ and then, with more time and planning, redirect the resolvers to micro-services. This is particularly useful for small companies, which are still testing ideas around. During the migration, back-end might change a lot, but front-end will be able to continue requesting the same stuff.
+
+As teams develop micro-services, a strategy is required to expose and change the public facing GraphQL API. There are some known strategies for it:
+
+### Remote schema stitching
+
+The gateway will get the schema from other GraphQL services and then stitch them together as the public-facing schema. This allows more freedom for teams, but it will disperse the API schema through multiple places, making it harder to test and easier to break, with a chance to have merge/stitching conflict.
+
+Although this can be a problem, teams can make releases without even touching the API Gateway, making the deploys isolated instead of requiring constant API Gateways deploys.
+
+> Recently, Apollo Server implemented [Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/), which will replace remote schema stitching. Bear in mind it is still Apollo specific.
+
+### API Gateway owns schemas
+
+Schema is contained locally on the gateway. This makes development and testing easier, as everything will be in the same place. The caveat is that the service will be constantly modified by multiple teams (resolvers, schemas), requiring constant gateway releases.
+
+As this service will be constantly modified by multiple people, rules around code formatting and style should be agreed through all teams to make the code uniform.
+
+> Personally, I prefer this one as it keeps everything in one place, gives high visibility on what is happening and allows easy discovery of other resolvers.
+
+### Combination of both above
+
+It is possible to mix both strategies, which is especially useful if the team wants to use remote schema stitching but have some services using GRPC or HTTP. Another way to tackle those non-GraphQL services is to put a server in front of it, allowing schema stitching and giving more flexibility for the responsible team.
+
+## 🙉 Caveats of using GraphQL
+
+### Services can get a bit chatty
```

Comment:
This might require one-two extra paragraphs to explain why this happens (GraphQL mechanics)